### ISSUE TITLE
Skip passed event and add descriptons

### ISSUE THREAD

**Issue by Cybersmurfen** [tybritten/ical-sensor-homeassistant#26]
As it is now, the component lists passed events if they are on the same day. As I use automations to check next event it is quite hard to switch between different sensors if the event is passed. Unfortunately I havent figured out the code for this.

So, if the end time for the event is passed, it is better to just let it fall out from the sensors list.

It would be a good idea to change the eta to minutes or seconds (Not very important)

Also - it is a good idea to also have the description (or notes) added:

I just added this to the code:

                if 'DESCRIPTION' in event:
                    event_dict['description'] = event['DESCRIPTION']

and

self._event_attributes = {
            'name': None,
            'location': None,
            'description': None,
            'start': None,
            'end': None,
            'eta': None
        }

and

            self._event_attributes['location'] = location
            description = val.get('description', '')

**Comment by Cybersmurfen** [comment]
I rewrote the sensor slightly:
Added description
Added to skip passed events the minute when they are passed
Added time_left as an attribute
Removed the date from the state
Added the prefix ical_ to the sensor
Sorted the events so that a recurring event is shown after an edited recurring event, depending on the time change
Changed the order of checking events, first normal ones, after that recurring - just to ensure the sensor is loaded.
Commented some issues and all changes in the code

**Comment by Cybersmurfen** [comment]
"""
Support for iCal-URLs

"""
import datetime as dt
import pytz
import logging
from datetime import timedelta
from dateutil.rrule import rrulestr, rruleset


import requests

from homeassistant.helpers.entity import Entity
from homeassistant.util import Throttle

_LOGGER = logging.getLogger(__name__)
REQUIREMENTS = ['icalendar', 'requests', 'arrow>=0.10.0']

VERSION = "0.6"
ICON = 'mdi:calendar'
DEFAULT_NAME = 'iCal Sensor'
DEFAULT_MAX_EVENTS = 5
PLATFORM = 'ical'
SCAN_INTERVAL = timedelta(minutes=2)

# Return cached results if last scan was less then this time ago.
MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=120)


def setup_platform(hass, config, add_entities, discovery_info=None):
    """Setup the iCal Sensor."""
    url = config.get('url')
    name = config.get('name', DEFAULT_NAME)
    maxevents = config.get('maxevents', DEFAULT_MAX_EVENTS)

    if url is None:
        _LOGGER.error('Missing required variable: "url"')
        return False

    data_object = ICalData(url)
    data_object.update()

    if data_object.data is None:
        _LOGGER.error('Unable to fetch iCal')
        return False

    sensors = []
    for eventnumber in range(maxevents):
        sensors.append(ICalSensor(hass, data_object, name, eventnumber))

    add_entities(sensors)


def dateparser(calendar, date):
    """
    Takes a calendar and a date, and returns a sorted list
    of events on or after that date.
    """
    import arrow
    events = []

    for event in calendar.walk('VEVENT'):
        # RRULEs turns out to be harder than initially thought.  
        # This is maiiony due to pythons handling of TZ-naive and TZ-aware timestamps, and the inconsistensies 
        # in the way RRULEs are implemented in the icalendar library.  
		# Changed so that the RRULE runs after the single events. RRULEs might timeout and if the sensor is initialized it can run anyways.
        if not 'RRULE' in event:
            if isinstance(event['DTSTART'].dt, dt.date):
                start = arrow.get(str(event['DTSTART'].dt))
            else:
                start = event['DTSTART'].dt

            # Add the end info if present.
            if 'DTEND' in event:
                if isinstance(event['DTEND'].dt, dt.date):
                    end = arrow.get(str(event['DTEND'].dt))
					
                else:
                    end = event['DTEND'].dt
            else:
                # Use "start" as end if no end is set
                end = start
			# Add startsort, so that recurring events always starts after modified events if they have the same start time.
            startsort = start
            # Noticed that somwhere the UTC difference was lost, so recalculating. Only purpose is to see if event has passed, same check in the RRULE loop
            utcal = (int(start.strftime("%z")[3:4])*600) + (int(start.strftime("%z")[1:3])*3600)
			
			# Add or remove seconds for ETA and Timeleft and then convert to minutes, Timeleft is the amount in minutes left of the event, same procedure in the RRULE loop
            if start.strftime("%z")[0:1] == '+':
                eta = (int(start.strftime("%s")) - int(date.strftime("%s")) - int(utcal)) / 60
                timeleft = (int(end.strftime("%s")) - int(date.strftime("%s")) - int(utcal)) / 60
            else:
                eta = (int(start.strftime("%s")) - int(date.strftime("%s")) + int(utcal)) / 60
                timeleft = (int(end.strftime("%s")) - int(date.strftime("%s")) + int(utcal)) / 60
			

            event_dict = {
                'name': event['SUMMARY'],
                'start': start,
			    'startsort': startsort,
                'end': end,
				'updated': date,
				'eta': eta,
				'timeleft': timeleft
            }

            # Add location if present
            if 'LOCATION' in event:
                event_dict['location'] = event['LOCATION']
					
			# Add description if present	
            if 'DESCRIPTION' in event:
                event_dict['description'] = event['DESCRIPTION']

            # If end time has passed, skip the event. Better to add an offset in config if we want old events kept #todo
            if int(timeleft) > 0:
                events.append(event_dict)

        else:
            start_rules = rruleset()
            end_rules = rruleset()
            rrule = event['RRULE']

            # Since we dont get both the start and the end in a single object, we need to generate two lists,
            # One of all the DTSTARTs and another list of all the DTENDs

            # Lets try to generate a list of all DTSTARTs.  
            # We just do our best, and will catch the exeption when it fails and continue to the next event.
            try:
                start_rules.rrule(rrulestr(rrule.to_ical().decode("utf-8"), dtstart = event['DTSTART'].dt))
            except Exception as e:
                _LOGGER.error(e)
                continue

            # If DTEND is not defined, this will fail. 
            # In that case, we will just use the DTSTARTs as DTENDs
            try:
                end_rules.rrule(rrulestr(rrule.to_ical().decode("utf-8"), dtstart = event['DTEND'].dt))
            except Exception as e:
                end_rules = start_rules

            # EXDATEs are hard to parse.  They might be a list, or just a single object.
            # They might contain TZ-data, they might not...
            # We just do our best, and will catch the exeption when it fails and move on the the next event.
            try:
                if 'EXDATE' in event:
                    if isinstance(event['EXDATE'], list):
                        for exdate in event['EXDATE']:
                            for edate in exdate.dts:
                                start_rules.exdate(edate.dt)
                                end_rules.exdate(edate.dt)
                    else:
                        for edate in event['EXDATE'].dts:
                            start_rules.exdate(edate.dt)
                            end_rules.exdate(edate.dt)
            except Exception as e:
                _LOGGER.error(e)
                continue

            # UNTIL will probably contain a TZ.  But if UNTIL is not defined, the RRULE seems to
            # usually be TZ-naive.  So we defined "now" as either TZ-aware or naive based on the 
            # presence of a UNTIL-tag.  Probably not perfect, but seems to work "most of the time"
            if 'UNTIL' in rrule:
                now = date.datetime
            else:
                now = dt.datetime.now()

            # Lets get all RRULE-generated events which will start 183 days before today and end 183 days after today
            # to ensure we are catching recurring events that might already have started.
            try:
                starts = start_rules.between(after=(now - timedelta(days=183)), before=(now + timedelta(days=183)))
                ends = end_rules.between(after=(now - timedelta(days=183)), before=(now + timedelta(days=183)))
            except Exception as e:
                _LOGGER.info(e)
                continue

            # We might get RRULEs that does not fall within the limits above, lets just skip them
            if len(starts) < 1:
                continue

            # It has to be a better way to do this...But at least it seems to work for now.
            ends.reverse()
            for start in starts:
                # Sometimes we dont get the same number of starts and ends...
                if len(ends) == 0:
                    continue
                end = arrow.get(str(ends.pop()))
                start = arrow.get(str(start))
                startsort = arrow.get(str(start + timedelta(seconds=1)))
				
                utcal = (int(start.strftime("%z")[3:4])*600) + (int(start.strftime("%z")[1:3])*3600)
			
                if start.strftime("%z")[0:1] == '+':
                    eta = (int(start.strftime("%s")) - int(date.strftime("%s")) - int(utcal)) / 60
                    timeleft = (int(end.strftime("%s")) - int(date.strftime("%s")) - int(utcal)) / 60
                else:
                    eta = (int(start.strftime("%s")) - int(date.strftime("%s")) + int(utcal)) / 60
                    timeleft = (int(end.strftime("%s")) - int(date.strftime("%s")) + int(utcal)) / 60
				
				
                # We should now have both a start and end arrow for the same event
                event_dict = {
                    'name': event['SUMMARY'],
                    'start': start,
				    'startsort': startsort,
                    'end': end,
					'updated': date,
					'eta': eta,
					'timeleft': timeleft
                }

                # Add location if present
                if 'LOCATION' in event:
                    event_dict['location'] = event['LOCATION']
					
				# Add description if present
                if 'DESCRIPTION' in event:
                    event_dict['description'] = event['DESCRIPTION']
                
				# If end time has passed, skip the event. Better to add an offset in config if we want old events kept #todo
                if int(timeleft) > 0:
                    events.append(event_dict)

    sorted_events = sorted(events, key=lambda k: k['startsort'])
    _LOGGER.debug(sorted_events)
    return sorted_events


# pylint: disable=too-few-public-methods
class ICalSensor(Entity):
    """
    Implementation of a iCal sensor.
    Represents the Nth upcoming event.
    May have a name like 'sensor.mycalander_event_0' for the first
    upcoming event.
    """
    def __init__(self, hass, data_object, sensor_name, eventnumber):
        """
        Initialize the sensor.
        sensor_name is typically the name of the calendar.
        eventnumber indicates which upcoming event this is, starting at zero
        """
        self._eventno = eventnumber
        self._hass = hass
        self.data_object = data_object
        self._name = 'ical_' + sensor_name + '_event_' + str(eventnumber)
        self._event_attributes = {}
        self.update()

    @property
    def name(self):
        """Return the name of the sensor."""
        return self._name

    @property
    def icon(self):
        """Return the icon for the frontend."""
        return ICON

    @property
    def state(self):
        """Return the date of the next event."""
        return self._state

    @property
    def device_state_attributes(self):
        """The name and date of the event."""
        return self._event_attributes

    def update(self):
        """Get the latest update and set the state and attributes."""
        # Defaults:
        self._state = "No event"
        # I guess the number and details of attributes probably
        # shouldn't change, so we should really prepopulate them.
        self._event_attributes = {
            'name': None,
            'location': None,
            'description': None,
            'start': None,
            'end': None,
            'updated': None,
            'eta': None,		
            'timeleft': None
        }
        # Get the data
        self.data_object.update()

        event_list = self.data_object.data
        if event_list and (self._eventno < len(event_list)):
            val = event_list[self._eventno]
            start = val['start'].datetime
            self._event_attributes['start'] = start
            end = val['end'].datetime
            self._event_attributes['end'] = end
            updated = val['updated'].datetime
            self._event_attributes['updated'] = updated
            location = val.get('location', '')
            self._event_attributes['location'] = location
            description = val.get('description', '')
            self._event_attributes['description'] = description
            eta = int(round(val['eta']))
            self._event_attributes['eta'] = eta
            timeleft = int(round(val['timeleft']))
            self._event_attributes['timeleft'] = timeleft
            name = val.get('name', 'unknown')
            self._event_attributes['name'] = name
            self._state = name
			# Issue: If the recurring calendar is edited with a single event change, the single event is displayed as event #1 and the original event as #0
			# There should be a comparision and skip the edited recurring event. Temporary solution is to simply add one second to all recurring events
			# Issue 2: Ongoing recurring events are not shown, probably sorted out somewhere.

# pylint: disable=too-few-public-methods
class ICalData(object):
    """
    Class for handling the data retrieval.
    The 'data' field is the sorted list of future events.
    """

    def __init__(self, resource):
        self._request = requests.Request('GET', resource).prepare()
        self.data = None

    @Throttle(MIN_TIME_BETWEEN_UPDATES)
    def update(self):
        import arrow
        import icalendar

        self.data = []

        try:
            with requests.Session() as sess:
                response = sess.send(self._request, timeout=90)

            cal = icalendar.Calendar.from_ical(response.text.replace("\x00", ""))
            today = arrow.utcnow()
            events = dateparser(cal, today)

            self.data = events

        except requests.exceptions.RequestException as e:
            _LOGGER.error("Error fetching data: %s", self._request)
            _LOGGER.error(e)
            self.data = None

**Comment by Olen** [comment]
Hi, and thanks for the updates.
Could you please create a pull request with the different fixes?
Also - I am working on a major rewrite of recurring events - see Issue #23 (last message).
I have not merget these changes yet, but they seem to work much better for recurring events.

**Comment by Cybersmurfen** [comment]
Well, I'm not totally satisified with it. The recurring events are quite odd. If I add a recurring event starting at a saturday night and stopping at sunday morning, the stop date of each event is the next sunday - i.e. 8 days instead of 1 day later.

The solution would be to check the start and stop, calculate the time between in minutes and then calculate the end time of the recurring events.

But if I edit a recurring event, the original one is still there (Of course it is, it's normal), but should be overwritten by the edited one.

So you could see my code as a quick fix, but not solving the RRULE issue, other than changing the order, but if I edit a recurring event to start like 15 minutes later this doesn't work.
